### PR TITLE
[DeviceSaniitizer] Force outline call for setting private shadow memory

### PR DESCRIPTION
**Pull request by zhaomaosu**
By default, address sanitizer will inline call for setting private shadow memory with small size. However, if work group size is too large, the private shadow memory may allocate failed. We need to check if shadow base is null before trying to poison it.

### CODE REVIEW

**Review by yingcong-wu — APPROVED** [review]
lgtm, but let's wait for another approval.

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (AllanZyne) [inline llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:3694]: maybe we can add "return" here to simplify the logic?

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (AllanZyne) [inline llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:4085]: no need

Comment (zhaomaosu) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: Removed

**Review by AllanZyne — APPROVED** [review]

File: llvm/test/Instrumentation/AddressSanitizer/SPIRV/instrument_private_address_space.ll
Comment (wenju-he) [inline llvm/test/Instrumentation/AddressSanitizer/SPIRV/instrument_private_address_space.ll:25]:
```suggestion
  ; CHECK: call void @__asan_set_shadow_private(i64 %5, i64 1, i8 -13)
```
for my education, why does the last parameter changes?

File: sycl/test-e2e/AddressSanitizer/out-of-bounds/USM/large_group_size.cpp
Comment (AllanZyne) [inline sycl/test-e2e/AddressSanitizer/out-of-bounds/USM/large_group_size.cpp:1]: Our tests can run on gpu now.

```suggestion
// REQUIRES: linux
```

File: llvm/test/Instrumentation/AddressSanitizer/SPIRV/instrument_private_address_space.ll
Comment (zhaomaosu) [reply llvm/test/Instrumentation/AddressSanitizer/SPIRV/instrument_private_address_space.ll:25]: Hi @wenju-he, the first call is to poison left red zone, and -15 is magic number of left red zone. And second call is to poison right red zone, -13 is magic number of it.

Comment (wenju-he) [reply llvm/test/Instrumentation/AddressSanitizer/SPIRV/instrument_private_address_space.ll:25]: thank you @zhaomaosu

**Comment by zhaomaosu** [comment]
Hi @intel/llvm-gatekeepers, this PR is ready to be merged. And pre-ci failures are unrelated to this change.

**Comment by zhaomaosu** [comment]
Hi @intel/llvm-gatekeepers, could you please help merge this PR? Thanks.

**Comment by sommerlukas** [comment]
@zhaomaosu This PR failed in post-commit, can you take a look if the failure is related to your changes: 
https://github.com/intel/llvm/actions/runs/10175954748/job/28144955957

Specifically:
```
Timed Out Tests (1):
  SYCL :: Plugin/level_zero_usm_residency.cpp
```

**Comment by zhaomaosu** [comment]
Hi @sommerlukas, I believe this failure is unrelated to my change. My change only affect device sanitizer. And the failed test doesn't use device sanitizer.

**Comment by sommerlukas** [comment]
Reported the unrelated failure in https://github.com/intel/llvm/issues/14863

**Pull request merged**
